napari/layers/image/_image_base_mouse_bindings.py

```diff
@@ -0,0 +1,119 @@
+import numpy as np
+
+from ...utils.geometry import point_in_bounding_box
+
+
+def on_embedded_plane_drag(layer, event):
+    """Shift a rendered plane along its normal vector.
+
+    This function will shift a plane along its normal vector when the plane is
+    clicked and dragged. The general strategy is to
+    1) find both the plane normal vector and the mouse drag vector in canvas
+    coordinates
+    2) calculate how far to move the plane in canvas coordinates, this is done
+    by projecting the mouse drag vector onto the (normalised) plane normal
+    vector
+    3) transform this drag distance (canvas coordinates) into data coordinates
+    4) update the plane position
+    """
+    # local import to avoid circular import
+    from ...viewer import current_viewer
+
+    if not layer.embedded_plane.enabled:  # early exit clause
+        return
+
+    # Get a reference to the viewer and store current drag state
+    viewer = current_viewer()
+    dragged = False
+
+    # Calculate intersection of click with data bounding box
+    near_point, far_point = layer.get_ray_intersections(
+        event.position,
+        event.view_direction,
+        event.dims_displayed,
+    )
+
+    # exit if click is outside data bounding box
+    if near_point is None and far_point is None:
+        return
+
+    # Calculate intersection of click with plane through data
+    intersection = layer.embedded_plane.intersect_with_line(
+        line_position=near_point, line_direction=event.view_direction
+    )
+
+    # Check if click was on plane by checking if intersection occurs within
+    # data bounding box. If so, exit early.
+    if not point_in_bounding_box(intersection, layer.extent.data):
+        return
+
+    # Get transform which maps from data (vispy) to canvas
+    visual2canvas = viewer.window.qt_viewer.layer_to_visual[
+        layer
+    ].node.get_transform(map_from="visual", map_to="canvas")
+
+    # Get plane parameters in vispy coordinates (zyx -> xyz)
+    plane_normal_data_vispy = np.array(layer.embedded_plane.normal)[[2, 1, 0]]
+    plane_position_data_vispy = np.array(layer.embedded_plane.position)[
+        [2, 1, 0]
+    ]
+
+    # Find start and end positions of plane normal in canvas coordinates
+    plane_normal_start_canvas = visual2canvas.map(plane_position_data_vispy)
+    plane_normal_end_canvas = visual2canvas.map(
+        plane_position_data_vispy + plane_normal_data_vispy
+    )
```

Comment:
Looks like you need the `visual2canvas` transform and to do this mapping here. Is there nothing on the mouse event or layer that provides this information? Maybe @kevinyamauchi can look too, but I'd really prefer to find another way to do this as we've now leaked even more of vispy into the layer

Reply:
yeah, I definitely agree here. since you're ultimately just using the viewer to retrieve an object (a transform) related to the layer in the callback, using the current viewer to get right back to an attribute of the layer feels like a big workaround. (this would not be a case that I would argue for in the discussion happening at #3147).

If that transform can't be retrieved directly on the layer.transforms... I'd be curious to hear if it could be

Reply:
You're exactly right - that transform isn't available on the layer but tucked away in the vispy layer. 

This transform could in theory be put in the layer transforms... @kevinyamauchi and I have talked about this and think the way to go is to do these calculations in world/data coordinates and avoid using the canvas coordinate system